bindings/matrix-sdk-ffi/src/event_filter.rs

```diff
@@ -0,0 +1,198 @@
+use matrix_sdk_ui::timeline::event_type_filter::TimelineEventTypeFilter as InnerTimelineEventTypeFilter;
+use ruma::events::{AnySyncTimelineEvent, TimelineEventType};
+
+#[derive(uniffi::Enum)]
+pub enum TimelineEventTypeFilter {
+    Include { event_types: Vec<FilterTimelineEventType> },
+    Exclude { event_types: Vec<FilterTimelineEventType> },
+}
+
+impl TimelineEventTypeFilter {
+    pub(crate) fn filter(&self, event: &AnySyncTimelineEvent) -> bool {
+        match self {
+            TimelineEventTypeFilter::Include { event_types } => {
+                let event_types: Vec<TimelineEventType> =
+                    event_types.iter().map(|t| t.clone().into()).collect();
+                InnerTimelineEventTypeFilter::Include(event_types).filter(event)
+            }
+            TimelineEventTypeFilter::Exclude { event_types } => {
+                let event_types: Vec<TimelineEventType> =
+                    event_types.iter().map(|t| t.clone().into()).collect();
+                InnerTimelineEventTypeFilter::Exclude(event_types).filter(event)
+            }
+        }
+    }
+}
+
+#[derive(uniffi::Enum, Clone)]
+pub enum FilterTimelineEventType {
+    MessageLike { event_type: FilterMessageLikeEventType },
+    State { event_type: FilterStateEventType },
+}
```

Comment:
Instead of this new family of types, can we use the existing FFI type `TimelineEventType`? Otherwise it's hard to ensure that every time a new event type is added to Ruma, we'll not forget about adding it to multiple places at the FFI layer.

Reply:
I don't think we can reuse it. ffi's `TimelineEventType` must contain either a `MessageLikeEventContent` or `StateEventContent` and they're not just enums with simple type values, it also contains the actual content of the event. So we can't say `TimelineEventType.MessageLike(MessageLikeEventContent::RoomMessage { .. }` in Kotlin or Swift, we'd have to specify the whole actual content values if I'm not mistaken.

Reply:
Is creating these enum cases something we could solve with a macro? Not that I know how to write macros, but it seems like something that could be automated.

Reply:
Ruma does have [a public `TimelineEventType`](https://docs.rs/ruma/latest/ruma/events/enum.TimelineEventType.html) that doesn't hold any data other than the event type :eyes: Can we use that instead?

(The alternative being, move both enums to the sdk crate, properly document and unit-test them in isolation. For what it's worth: these days we can annotate sdk's enums with `#[derive(uniffi::Enum)]` conditionally, to avoid the FFI copy of the enum.)

Reply:
Can we expose ruma types in the FFI layer? Also, do you have any examples of the conditional annotation?

Reply:
Doh, you're right, we'd need to expose them (and we'd need to update Ruma's code to also have that conditional uniffi feature + derive macro — how does that sound @jplatte?).

An example of the conditional derive annotation would be `matrix_sdk_ui::timeline::pagination::BackPaginationStatus`.